2.x posts api

acobster opened the pull request:
Work in progress final port of the Posts API to use Factories under the hood.

- [x] Switch `::get_post[s]()` API to use Factories
- [x] Replace a bunch of direct instantiations in the tests
- [x] Fix `Term::posts()`
- [x] `Post::build()` method
- [x] Attachment/Image class map callback
- [x] Update Twig functions (@gchtr doing on #2178)
- [x] Simplify `PostQuery::__construct()` to only accepting a `WP_Query` instance (with no wrapping it in an array)
- [x] Fix #2178 and merge in to this feature branch
- [x] Make `Post::__construct()` protected, annotate `WP_Post` arg
- [x] `Timber::get_attachment_by()`
- [x] Close the loop on whether we need to pass an array as a second arg to `wp_check_filetype()`
- [x] `Attachment::__construct()` has an option to pass an array like `['ID' => 123]` ~...are we good to drop support for this? Otherwise we would need to handle this special case in `::get_post()`.~ we need to support this for ACF integration, so add a special case for this in `::get_post()` (or `PostFactory`?)
- [x] `@expectedIncorrectUsage` is failing randomly on those two tests...y tho.
- [x] deprecate `Attachment::get_pathinfo()` and `Image::get_pathinfo()` and use `Attachment::pathinfo()` only to be in line with other post method names.
- [x] Support the `merge_default` option in `::get_post[s]()`

gchtr commented:
@acobster Just a reminder that there’s also #2145, where I added some helpful `Helper::doing_it_wrong()` calls as well as DocBlocks for the new `Timber::get_post()` and `Timber::get_posts()` API. I don’t mind if you just copy what you need and then close my pull request, because that might be easier than merging.

acobster commented:
@gchtr thanks! I pulled most of that stuff in - see latest updates.

I have a couple design questions for you and @jarednova. We need to figure out a couple things around Images/Attachments:

* `PostFactory` has some built-in logic to return an `Attachment` for the attachment post type, but I'm wondering if there's a way to cleanly discern that we should return the more specific `Image` class? This could be accomplished simply by setting a callback function for that post type, but what I don't know is what it should check...
* I'm not sure what to do about instantiating images from files/URLs. There are a bunch of tests that do stuff like `new Timber\Image($url)`. What should replace that? Maybe a static `Image::from( ... )` method?

jarednova commented:
**PostFactory has some built-in logic to return an Attachment for the attachment post type, but I'm wondering if there's a way to cleanly discern that we should return the more specific Image class? This could be accomplished simply by setting a callback function for that post type, but what I don't know is what it should check...**
Hmmmm.... there's an `Image::is_image()` function we could use to share a common definition (and be extensible/filterable). Right now it counts these as recognized image extensions: `array( 'jpg', 'jpeg', 'jpe', 'gif', 'png' )` as images (`wepb` should probably be added). 

**I'm not sure what to do about instantiating images from files/URLs. There are a bunch of tests that do stuff like new Timber\Image($url). What should replace that? Maybe a static Image::from( ... ) method?**
I feel like what's going on when handling an external URL is really distinct from when things are pulled from the WP database. What about `Image::from_url(..)`?

acobster commented:
> there's an Image::is_image() function we could use to share a common definition

Gotcha! Only issue with this is that it's a protected instance method. I'll reference that implementation and make the extensions filterable. If folks want even more fine-grained control they can always implement their own Class Map callback for `attachment`.

> I feel like what's going on when handling an external URL is really distinct from when things are pulled from the WP database. What about `Image::from_url(..)`?

Agreed. Sounds good!

acobster commented:
@gchtr @jarednova there's some stuff left to do (see description) but please take a look at how it's going so far. In particular I skipped **all** the tests in `test-timber-post-getter.php` and ported tests for the functionality we're keeping over to `test-timber.php`. I could especially use eyes on that part to make sure I didn't unintentionally drop anything. Thanks!

acobster commented:
Also @gchtr @jarednova - is the Twig API documented [here](https://github.com/timber/timber/blob/2.x-docs-api/docs/v2/upgrade-guides/2.0.md#twig) what we landed on? I just want to make sure because I thought I remembered at some point discussing the possibility of e.g. `{{ TimberPost(...) }}` translating over to `{{ get_post(...) }}`. That may have been an early iteration or I may just be making things up again. :stuck_out_tongue:

gchtr commented:
> Also @gchtr @jarednova - is the Twig API documented [here](https://github.com/timber/timber/blob/2.x-docs-api/docs/v2/upgrade-guides/2.0.md#twig) what we landed on? I just want to make sure because I thought I remembered at some point discussing the possibility of e.g. `{{ TimberPost(...) }}` translating over to `{{ get_post(...) }}`. That may have been an early iteration or I may just be making things up again. :stuck_out_tongue: 

@acobster I worked on the API part and the new `{{ get_*(...) }}`  functions in Twig in https://github.com/timber/timber/pull/2178. But I think it's not yet documented in https://github.com/timber/timber/blob/2.x-docs-api/docs/v2/upgrade-guides/2.0.md.

I can have a look at the rest next week.

acobster commented:
@gchtr cool, I made some small changes over here mostly just to see if the tests would pass, but I'll probably revert it to avoid conflicting with your stuff on #2178. I will continue leaning on skipping tests temporarily for known issues, so the tests can be a kind of todo-list once both PRs are in `2.x-factories`. Sound good?

jarednova commented:
@acobster that sounds good! Similarly I think that tests and structure for `TermGetter` will have to be similarly migrated or removed

acobster commented:
@jarednova over on #2178 @gchtr mentions maybe needing `{{ get_image() }}` and/or `{{ get_attachment() }}`. I think we're going to need one or the other, because `PostFactory::from()` returns `false` if passed a non-numeric string. As discussed, this use-case is different enough that it deserves its own `Attachment::from_url()` method, and by that logic I don't want to add stuff in `PostFactory` to deal with this because then we'd have to distinguish between files, URLs, slugs, titles, and any other bogus strings. So my idea is to have a `Timber::get_attachment()` that assumes a non-numeric string is a file path or URL and defers to `::get_post()` for all other cases. We _could_ add a `::get_image()` too but I don't think this is necessary since it would probably just be an alias for `::get_attachment()`.

Now, regarding how that method would actually _work_...I still haven't quite wrapped my brain around how getting an `Attachment` instance from a file/URL works in 1.x but IIRC sometimes it doesn't end up associating it with a post: instead it's just a dumb wrapper around an image URL. (For that reason I've mostly avoided using that feature because it means you can't get different crops with `attachment.src('my-crop-size')`.) But I think we can change that, so here is what I'm picturing...

* Using a query like [this one](https://exceptionshub.com/get-attachment-id-by-file-path-in-wordpress.html), we try to fetch an attachment ID by its URL.
* If we don't find one, we might have a file path instead of a URL, so use `PathHelper` to convert to a URL and run the same query on the result of that conversion.
* If we manage to find an ID in either of the first two steps, we can call `PostFactory::from()` which will instantiate an `Attachment` or `Image` polymorphically.
* If we still don't have an ID we can return a dumb wrapper around the URL, like 1.x does. I'm not sure if it's useful to distinguish between `Attachment` and `Image` in this case but since it's only based on the filename anyway it's probably best for consistency.

Thoughts?

acobster commented:
> @acobster that sounds good! Similarly I think that tests and structure for TermGetter will have to be similarly migrated or removed

@jarednova good point. Because Terms are much simpler, I feel a lot more confident in the test coverage that's already there, but can you go over the `TermGetter` tests and see if there's anything we should migrate?

gchtr commented:
> So my idea is to have a `Timber::get_attachment()` that assumes a non-numeric string is a file path or URL and defers to `::get_post()` for all other cases. We _could_ add a `::get_image()` too but I don't think this is necessary since it would probably just be an alias for `::get_attachment()`.

You’re right, that’s a good reason to have `Timber::get_attachment()`! But as you say yourself, `Timber::get_attachment( 'https://example.org/image.jpg' )` won’t actually work with a WordPress attachment, but only load the image and create an object to work with.

I remember a discussion about using a separate class for these "interim" objects in https://github.com/timber/timber/pull/1788#pullrequestreview-155613762 and https://github.com/timber/timber/pull/1788#issuecomment-422521668.

And I think it would be a good solution to separate that use case and maybe also use a separate API for working with URLs and paths. The first idea that pops into my head is to use a function like `Timber::get_attachment_from()` which would accept a path or a URL.

```php
Timber::get_attachment_from( 'path', 'path/to/image' );
Timber::get_attachment_from( 'url', 'https://example.org/image.jpg' );

// or even
Timber::get_attachment_from( 'https://example.org/image.jpg' );
```

This would be somewhat in line with functions like `Timber::get_post_by()` and `Timber::get_user_by()` that handle special cases like the one we have here.

If we’d have a special class for this, say `ExternalAttachment`, we could also use

```php
Timber::get_external_attachment( 'path/to/image' );
Timber::get_external_attachment( 'https://example.org/image.jpg' );
```

We would have to find a good name for that, though.

>     * Using a query like [this one](https://exceptionshub.com/get-attachment-id-by-file-path-in-wordpress.html), we try to fetch an attachment ID by its URL.

Wouldn’t that be quite the performance impact? I feel like if we try to work too much magic here, things could become quite complex again, right?

It’s so good you’re catching stuff like this! 👍

gchtr commented on tests/test-timber-pagination.php:
```diff
@@ -123,7 +123,7 @@ function testDoubleEncodedPaginationUrl() {
 
 		$link = Timber::compile_string("{{ posts.pagination.next.link|e('esc_url') }}", array(
-			'posts' => new Timber\PostQuery(),
+			'posts' => new PostQuery($GLOBALS['wp_query']),
```
I’m just picking one your changes randomly… there are quite a few of those.

Is this an example for the following comment of yours?

> I will continue leaning on skipping tests temporarily for known issues, so the tests can be a kind of todo-list once both PRs are in 2.x-factories. Sound good?

Are usages of `new PostQuery()` something that still needs work or are you deliberately using it here to test specific functionality in `PostQuery`?

acobster replied:
I tried to go through and replace all these instances in the tests, except for the skipped ones where there are open questions. Generally I think we should test at the highest level possible.

gchtr reviewed (commented):
> I will continue leaning on skipping tests temporarily for known issues, so the tests can be a kind of todo-list once both PRs are in `2.x-factories`. Sound good?

That sounds good 👍 . I’ll have to closely look at the tests again. More questions directly in the code.

I like the new `timber/post/import_data` filter!

acobster commented:
```php
Timber::get_attachment_from( 'path', 'path/to/image' );
Timber::get_attachment_from( 'url', 'https://example.org/image.jpg' );

// or even
Timber::get_attachment_from( 'https://example.org/image.jpg' );
```

So is the idea that if it got passed one arg, it would do the URL/Path cascade query I described above? I like that a lot! Should probably be `::get_attachment_by()` for consistency though. :)

Just to be clear - that would still mean there's no single method in Twig or otherwise that could handle **either** an ID/`WP_Post` **or** a path/URL? That's consistent with the `::get_post()` and `::get_post_by()` distinction, just confirming.

> If we’d have a special class for this, say ExternalAttachment, we could also use ... `Timber::get_external_attachment( 'path/to/image' );`

I'm a little hesitant to do this, just because I'm not sure how useful it'd be. If it's a truly external URL that WP knows nothing about, how many of the methods actually add value beyond the URL, which we already have? Like I said I avoided using that in the past so I'm not sure if there are benefits I'm not aware of.

> Wouldn’t that be quite the performance impact? I feel like if we try to work too much magic here, things could become quite complex again, right?

Yes, which is part of why it should be separate. We should document the performance implications and encourage users to always pass a `WP_Post` or attachment ID directly to {{ get_post() }} if they can.

acobster commented on lib/Timber.php:
```diff
@@ -632,10 +775,9 @@ public static function context() {
 		if ( is_singular() ) {
-			$post = ( new Post() )->setup();
-			$context['post'] = $post;
+			$context['post'] = Timber::get_post()->setup();
```
Sounds good!

gchtr commented:
> So is the idea that if it got passed one arg, it would do the URL/Path cascade query I described above? I like that a lot!

I don’t know whether both patterns should be support or only one of them. If it’s easy enough to work with just a string, then let’s go for it!

> Should probably be `::get_attachment_by()` for consistency though. :)

Yes. My only thinking was that it probably isn’t the same pattern when passing an external URL, but you could also pass an attachment URL and get the usual `Attachment` post, so we’d have consistency there 👍.

> Just to be clear - that would still mean there's no single method in Twig or otherwise that could handle **either** an ID/`WP_Post` **or** a path/URL? That's consistent with the `::get_post()` and `::get_post_by()` distinction, just confirming.

Yes, I think so.

> I'm a little hesitant to do this, just because I'm not sure how useful it'd be. If it's a truly external URL that WP knows nothing about, how many of the methods actually add value beyond the URL, which we already have? Like I said I avoided using that in the past so I'm not sure if there are benefits I'm not aware of.

I have no experience with this at all, so I’m glad you can chip in with some insights. I just remembered that we already considered this and wanted to throw it in. Let’s not do it then.

> Yes, which is part of why it should be separate. We should document the performance implications and encourage users to always pass a `WP_Post` or attachment ID directly to {{ get_post() }} if they can.

So, you would still support passing an URL or a path to `Timber::get_post()`, but the recommended way would be to use `Timber::get_attachment_by()`? What if we wouldn’t support it in `Timber::get_post()` and only support it via `Timber::get_attachment_by()`?

acobster commented:
> So, you would still support passing an URL or a path to `Timber::get_post()`, but the recommended way would be to use `Timber::get_attachment_by()`? What if we wouldn’t support it in `Timber::get_post()` and only support it via `Timber::get_attachment_by()`?

No, the second one is what I mean. :) Like I said before, no single method should support everything under the sun. All I'm saying here is that you have the option of starting with a URL if that's all you have, but if you have something more specific like an attachment ID, you should use that because it's more performant.

> I don’t know whether both patterns should be support or only one of them. If it’s easy enough to work with just a string, then let’s go for it!

The only reason I can see for also supporting arity-2 is if you know you have a path, and you want to skip the initial db round-trip for the URL check. So:

```php
$path = "/path/to/file.gif";

// This is slower because it tries a useless query by URL
$attachment = Timber::get_attachment_by( $path );

// This gives us the same post, but it's faster because it skips the URL query
$attachment = Timber::get_attachment_by( "path", $path );
```

gchtr commented:
> > So, you would still support passing an URL or a path to `Timber::get_post()`, but the recommended way would be to use `Timber::get_attachment_by()`? What if we wouldn’t support it in `Timber::get_post()` and only support it via `Timber::get_attachment_by()`?
> 
> No, the second one is what I mean. :) Like I said before, no single method should support everything under the sun. All I'm saying here is that you have the option of starting with a URL if that's all you have, but if you have something more specific like an attachment ID, you should use that because it's more performant.

So then we’re on the same page 👍 

> The only reason I can see for also supporting arity-2 is if you know you have a path, and you want to skip the initial db round-trip for the URL check.

And it would be consistent with the other `get_*_by()` functions. Can we assume that developers always know whether they have a path or a URL?

acobster commented:
> Can we assume that developers always know whether they have a path or a URL?

There might be cases where you could be dealing with either one, which is why I think it'd be nice to support either arity. :)

acobster commented:
OK, I've implemented that. For some reason the tests are failing to honor the `@expectedIncorrectUsage` tag...

```
$ vendor/bin/phpunit tests/test-timber-attachment.php 
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
Use http://build.starter-theme.dev/ for testing with UI
PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

Error:         No code coverage driver is available

S.....EE....SSS.SSSS.....                                         25 / 25 (100%)

Time: 7.83 seconds, Memory: 36.25 MB

There were 2 errors:

1) TestTimberAttachment::testGetAttachmentByUrlDoingItWrong
[ Timber ] Timber::get_attachment_by() was called <strong>incorrectly</strong>. Passing "url" as the first arg requires passing a URL as the second arg. Please see Debugging in WordPress (https://wordpress.org/support/article/debugging-in-wordpress/) as well as Debugging in Timber (https://timber.github.io/docs/guides/debugging/) for more information. (This message was added in Timber version 2.0.0.)

/srv/www/timber/lib/Helper.php:355
/srv/www/timber/lib/Timber.php:459
/srv/www/timber/tests/test-timber-attachment.php:78

2) TestTimberAttachment::testGetAttachmentByPathDoingItWrong
[ Timber ] Timber::get_attachment_by() was called <strong>incorrectly</strong>. Passing "path" as the first arg requires passing an absolute path as the second arg. Please see Debugging in WordPress (https://wordpress.org/support/article/debugging-in-wordpress/) as well as Debugging in Timber (https://timber.github.io/docs/guides/debugging/) for more information. (This message was added in Timber version 2.0.0.)

/srv/www/timber/lib/Helper.php:355
/srv/www/timber/lib/Timber.php:475
/srv/www/timber/tests/test-timber-attachment.php:85

ERRORS!
Tests: 25, Assertions: 33, Errors: 2, Skipped: 8.
```

No idea why. Will revisit tomorrow.

acobster commented on lib/Timber.php:
```diff
@@ -129,37 +174,130 @@ class_alias( 'Timber\Timber', 'Timber' );
+				'2.0.0'
+			);
+
+			$options = [];
```
Wasn't sure about this. Do we want to support passing a query to `::get_post()` (singular)? If so, we should also support `$options['merge_default']`.

acobster replied:
cc @jarednova

gchtr replied:
> Do we want to support passing a query to `::get_post()` (singular)? If so, we should also support `$options['merge_default']`.

Oh, now I see what that’s about! I’m not sure yet, either. When I think about what the equivalent with `Timber::get_posts()` would be, I think supporting queries for `Timber::get_post()` would be quite nice.

Would we have to add `'posts_per_page' => 1` to the query, if it isn’t set?

I would definitely have to add a small section in #2073 for this.

acobster commented on lib/Factory/PostFactory.php:
```diff
@@ -71,31 +73,48 @@ protected function from_wp_query(WP_Query $query) : Iterable {
-  protected function build(WP_Post $post) : CoreInterface {
+	protected function is_image(WP_Post $post) {
+		$src   = wp_get_attachment_url( $post->ID );
+		$check = wp_check_filetype( PathHelper::basename( $src ) );
```
If we do that I think we'd also need to change the `timber/post/image_extensions` filter to not only care about the extensions but also about the mime type, so users retain control over which attachments are considered images or not. So rather than a flat array of extensions, we'd run the filter earlier and usage would look like:

```php
add_filter('timber/post/image_filetypes', function(array $ext_regex_to_mime) {
  return array_merge($ext_regex_to_mime, [
    '/ext/' => 'image/something',
  ]);
});
```

This is more mental load on the user so I'm curious what you're concerned about here. :) Security? Performance?

acobster replied:
For context, I was going off of what @jarednova said in an earlier comment:

> there's an Image::is_image() function we could use to share a common definition (and be extensible/filterable)

Hence the need for an extra filter. That said, WP hard-codes this list in [`wp_attachment_is()`](https://developer.wordpress.org/reference/functions/wp_attachment_is/). So we could side-step this whole issue and just use that. It's not my favorite option because it doesn't support WEBP but just throwing that out there.

gchtr replied:
@acobster I was concerned about performance. On a site I recently worked on, [Blackfire](https://blackfire.io) showed that `wp_check_filetype()` was eating performance when I tried to load a lot of images on one page. I figured that when I pass the mime types I want to check for, then the `get_allowed_mime_types()` function inside `wp_check_filetype()` function as well as the `upload_mimes` filter are skipped. There was probably something else hooking into `upload_mimes` that used a lot of performance. But I thought since I’m only checking for images, I could do it like this:

```php
$mime_types = [
    'jpg|jpeg|jpe' => 'image/jpeg',
    'gif'          => 'image/gif',
    'png'          => 'image/png',
];

$check = wp_check_filetype( $filename, $mime_types );
```

But you are right, it gets trickier if we want to support WebP, because the WebP mime type needs to be added through the `upload_mimes` filter. If there should ever be a performance problem in the future, we can think about a solution then.

> For context, I was going off of what @jarednova said in an earlier comment:
> 
> > there's an Image::is_image() function we could use to share a common definition (and be extensible/filterable)

I’m all for that!

> That said, WP hard-codes this list in [`wp_attachment_is()`](https://developer.wordpress.org/reference/functions/wp_attachment_is/). So we could side-step this whole issue and just use that. It's not my favorite option because it doesn't support WEBP but just throwing that out there.

Yeah, not my favorite option either if it doesn’t support WebP.

acobster commented:
@gchtr somehow I started my own review instead of responding to your review comments directly, and now my responses are gone...thanks GH :roll_eyes: Anyway, just responded to your points. Also see outstanding questions in the issue description.

acobster commented:
@jarednova `Image::get_pathinfo()` was marked deprecated but `Attachment::get_pathinfo()` was not. Seemed like a simple oversight so I moved the deprecation notice up to `Attachment` and removed the `Image` one. Just FYI in case I'm missing something here.

gchtr commented:
Ok, about the outstanding points the in the description:

> - [ ] `Attachment::__construct()` has an option to pass an array like `['ID' => 123]` ...are we good to drop support for this? Otherwise we would need to handle this special case in `::get_post()`.

Can we add a `Helper::doing_it_wrong` for this?

> - [ ] Close the loop on whether we need to pass an array as a second arg to `wp_check_filetype()`

Let’s tick this off and **not pass** a second arg.

> @jarednova `Image::get_pathinfo()` was marked deprecated but `Attachment::get_pathinfo()` was not. Seemed like a simple oversight so I moved the deprecation notice up to `Attachment` and removed the `Image` one. Just FYI in case I'm missing something here.

Now it’s getting tricky. The `get_pathinfo()` function was deprecated with #2026 and #2042 in June 2019. However, we merged in #2072 in October 2019, which actually improves pathinfo support for files with special characters.

![](https://user-images.githubusercontent.com/2084481/92412902-06ad8680-f14e-11ea-89b0-645027902968.png)

So, I guess we should deprecate `Attachment::get_pathinfo()` and `Image::get_pathinfo()` and use `Attachment::pathinfo()` only to be in line with other post method names.

jarednova commented:
Can confirm, the missing deprecation for `Attachment::get_pathinfo()` is/was an oversight (not part of a grand secret plan, sadly)

acobster commented:
@gchtr @jarednova OK, great! I'd say this whole PR (and Factories in general!! :tada: :balloon: :partying_face:) is almost there - the biggest hurdle is going to be resolving merge conflicts in #2178 down the line, and bringing those updates up to speed with the decisions made here.

One [usage change](https://github.com/timber/timber/pull/2316/commits/ccb9947b8023702e285e1131c45981ff8eec4b59) that falls out of the lack of an `ExternalAttachment` class is that `User::avatar()` _has_ to return a URL (string) because we don't have anything to wrap it in. In the future it might be nice to have, say, `User::image()` return an Attachment, but for that we'd need to hash out resizing for Gravatars, which works differently from attachments. Consider:

```php
$gravatar = User::avatar(); // default-size gravatar URL (96px)
$gravatar_thumb = User::avatar(['size' => 150]); // 150x150 gravatar URL

// vs.

// These will no longer work - avatar is not an object
$oops = User::avatar()->src();
$nope = User::avatar()->src(['size' => 150]);
```

Are we okay with this breaking change or do we need to revisit a wrapper class? I'm personally in favor of dropping this for now and revisiting in a future 2.x release. I think we can support wrapping arbitrary URLs generically, but it needs more design work to make it compatible with Factories.

@gchtr let me know if you want me to jump on finalizing the changes in #2178 - I can make time this week. If you have time to do it, that's great too. :smile:

acobster commented:
@jarednova the other thing that breaks under the current Attachment design is getting an image from a retina URL:

```php
Timber::get_attachment_by( 'ur', 'http://example.org/wp-content/uploads/2020/09/eastern@1.5x.jpg' ); // -> false
```

There are a number of tests for this in `test-timber-image-retina.php`, e.g.:

```php
$compiled = Timber::compile_string($str, $data);
$img = Attachment::from_url($compiled); // [sic] - would be Timber::get_attachment_by( 'url', $compiled );
$this->assertContains('@3x', $compiled);
$this->assertEquals(300, $img->width());
```

Is this behavior we need to preserve, or is this an indirect way of testing `ImageHelper::retina_resize()`?

jarednova commented:
@acobster based on what I can see (in the `git blame` these are from 6 years ago so not gonna go by memory here) — they sure seem to just be validating the successful execution of `ImageHelper::retina_resize()`.  With the sample you provided ...

```php
Timber::get_attachment_by( 'url', 'http://example.org/wp-content/uploads/2020/09/eastern@1.5x.jpg' ); // -> false
```

... you're saying that trying to receive the primary attachment object based on providing the _retina_ URL (as opposed to the "regular") URL will return a false? I think this is fine. And if it turns out that it's not fine and creates an identifiable gap/burden — it's a discreet enough issue that can be addressed as an independent issue

acobster commented:
@jarednova exactly. :+1: 

BTW - any idea about why those (currently skipped) attachment tests are [failing](https://github.com/timber/timber/pull/2316#issuecomment-686186274)? I feel like I'm missing some PHPUnit or WP Unit Test trade secret...

jarednova commented:
Found it! In `TimberAttachment_UnitTestCase::setUp` we were missing a call to `parent::setUp();` which loads some of the tweaks on WordPress unit testing on top of PHPUnit (like the `@expectedIncorrectUsage` usage tag).

gchtr commented:
> @gchtr let me know if you want me to jump on finalizing the changes in #2178 - I can make time this week. If you have time to do it, that's great too. 😄

I’d be glad if you could finish this. You have a way better overview of what needs to be done to make this work with the factories. Thanks!

> Are we okay with this breaking change or do we need to revisit a wrapper class? I'm personally in favor of dropping this for now and revisiting in a future 2.x release. I think we can support wrapping arbitrary URLs generically, but it needs more design work to make it compatible with Factories.

Adding this as a breaking change and improving this in the future seams reasonable to me 👍 . Image handling is a different topic altogether, with looots of things to consider.

acobster commented:
> I’d be glad if you could finish this.

OK, done and merged in as part of this PR! They are inter-dependent, so one of them has to be done first. I figured I'd make that the smaller one.

acobster commented:
@gchtr @jarednova I'm calling this ready for review!

jarednova commented:
Excellent, awesome to hear @acobster — I'll add to my queue for the week!

jarednova commented on docs/v2/upgrade-guides/2.0.md:
```diff
@@ -303,7 +320,7 @@ The following functions were **removed from the codebase**, either because they
 - `get_field()` - use `{{ post.meta('my_field_name') }}` instead
 - `get_format()` – use `{{ post.format }}` instead
-- `get_image()` – use `{{ Image(post.my_image) }}` instead
+- `get_image()` – use `{{ get_post(post.my_image) }}` instead
```
```suggestion
- `get_image()` – use `{{ get_image(post.my_image) }}` instead
```

jarednova commented on lib/Attachment.php:
````diff
@@ -408,7 +393,7 @@ public function path() {
 	 * @example
 	 * ```twig
-	 * <a href="{{ Attachment(post.meta('job_pdf')).src }}" download>
+	 * <a href="{{ get_post(post.meta('job_pdf')).src }}" download>
````
```suggestion
	 * <a href="{{ get_attachment(post.meta('job_pdf')).src }}" download>
```

jarednova reviewed (changes requested):
@acobster we've already been through this, so I didn't see much else but a few docs things for you to check out. 

One last 🔧 /question is noticing the number of `Image($image_id)` that are now supposed to be called with `Timber::get_post($image_id)`. I'm wondering if we should have a `Timber::get_image()` (or `Timber::get_attachment()`) for these. Forgive me if this has already been discussed/resolved and my brain is just melting.

gchtr commented:
> One last 🔧 /question is noticing the number of `Image($image_id)` that are now supposed to be called with `Timber::get_post($image_id)`. I'm wondering if we should have a `Timber::get_image()` (or `Timber::get_attachment()`) for these. Forgive me if this has already been discussed/resolved and my brain is just melting. _(If we do this, let's mark as a `@todo` and tackle in a seperate PR so we can move forward w the merge here)_

@jarednova Yes, we discussed this in https://github.com/timber/timber/pull/2316#issuecomment-684004925 and previous as well as in https://github.com/timber/timber/pull/2178#issuecomment-583767717.

My take is: Using `{{ get_attachment() }}` and `{{ get_image() }}` as well as `Timber::get_attachment()` and `Timber::get_image()` in the code would help in understanding what you’re working with. We now have `Timber::get_attachment_by()` to handle the special cases, so I guess `Timber::get_attachment()` and `Timber::get_image()` would just be aliases for `Timber::get_post()`? What about `Timber::get_attachments()` and `Timber::get_images()`?

acobster commented on lib/Term.php:
```diff
@@ -586,61 +603,75 @@ public function path() {
-				'tax_query' => $default_tax_query,
-				'post_type' => $post_type_or_class
+
+		if ( is_int($query) ) {
```
OK, I will either merge that in or port it over here. Thanks!

acobster replied:
@gchtr to keep as much of your stuff as possible, I merged `2.x-posts-api` back into `2.x-term-posts` resolved merge conflicts favoring my changes on this branch, `2x.-posts-api`. Also added a note about passing an int for post count.

acobster commented:
I think as long as we say something like:

> Alias for `Timber::get_post()`. That means it's possible to get something other than an Attachment (such a plain Timber\Post object) back if you pass something other than an attachment ID. Intended to clarify your code when you already know you have an attachment ID.

...in the docs for `::get_(image|attachment)`, then @gchtr's proposal sounds good to me.

gchtr commented:
> I think as long as we say something like:
> 
> > Alias for `Timber::get_post()`. That means it's possible to get something other than an Attachment (such a plain Timber\Post object) back if you pass something other than an attachment ID. Intended to clarify your code when you already know you have an attachment ID.
> 
> ...in the docs for `::get_(image|attachment)`, then @gchtr's proposal sounds good to me.

Yes, sure!

acobster commented:
@gchtr @jarednova OK, I believe all the bookkeeping stuff for this PR. Are we good to go? :grimacing: :sweat_smile: 

![THE FINAL COUNTDOWN](https://media.giphy.com/media/RnX4q6yYDoYCI/giphy.gif)

jarednova commented:
Nothing but green checkmarks over here!
